### PR TITLE
v2 Migration guide

### PR DESCRIPTION
**Pull request by ctrlaltdylan**
A migration guide for v1 users that are being introduced to the new workflow builder.

### CODE REVIEW

File: docs/docs/migrate-from-v1/README.md [hunk 1]
```diff
@@ -0,0 +1,263 @@
+# Migrate from Pipedream v1
+
+The newest version of the Pipedream introduces features like creating code steps in [Python](/code/python), [Bash](/code/bash) and [Go](/code/go) from within the workflow builder.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:3]: "the Pipedream" -> "the Pipedream workflow builder"

File: docs/docs/migrate-from-v1/README.md [hunk 2]
```diff
@@ -0,0 +1,263 @@
+# Migrate from Pipedream v1
+
+The newest version of the Pipedream workflow introduces features like creating code steps in [Python](/code/python), [Bash](/code/bash) and [Go](/code/go).
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:3]: Still needs "builder" after "workflow"

File: docs/docs/migrate-from-v1/README.md [hunk 3]
```diff
@@ -0,0 +1,263 @@
+# Migrate from Pipedream v1
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:1]: I'd recommend using the language from Tod's Slack announcement here, and then add links to the relevant pages like Python, Go, etc. from there:

We have re-imagined the UX from the ground up, made the product much easier to use and have improved performance.  In addition, we are introducing powerful new features including:

- Improved builder to test and iterate without impacting live workflows
- Multi-language support including Node, Python, Golang & bash
- Granular testing including the ability to test individual steps and more
- Multiple triggers are now supported per workflow
- Improved forms for easier configuration and streamlined building

File: docs/docs/migrate-from-v1/README.md [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+The newest version of the Pipedream workflow introduces features like creating code steps in [Python](/code/python), [Bash](/code/bash) and [Go](/code/go).
+
+It also brings the [Node.js](/code/nodes) steps closer to our [Component API](/components/) which powers sources & actions in the platform.
+
+But this also means some of our features have changed. Here's what you need to know.
+
+## New Builder Overview
+
+Fundamentally, the new version of the workflow builder gives you the same abilities to build, test and deploy your workflows. However, there are some differences that change up your normal flow.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:11]: one way to change this:

> there are some differences that change up your normal flow.

from passive to active:

> you'll notice some differences in how to build workflows.

File: docs/docs/migrate-from-v1/README.md [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+### Editing vs Inspecting
+
+When you first open a deployed workflow, you're presented with the **inspector** version of the workflow. In this view you can see logs of past events, and select them to see the results of each step in the workflow.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:15]: Inspector should be capitalized.

I'd recommend noting how users switch between Inspect and Build / Edit mode, e.g. maybe talk about the ellipsis on the top-right corner and include a screenshot or something.

File: docs/docs/migrate-from-v1/README.md [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+To edit the workflow, click the **Edit** button in the top right hand corner. This will close the inspector and allow you to edit your workflow without the distraction of logs from the production flow.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:17]: Yeah re: the comment above, I guess this is the best place to say how to do it the other way: when you're in Build mode, click on the ellipsis in the top-right to get back to the Inspector

File: docs/docs/migrate-from-v1/README.md [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+### Testing Changes
+
+In the v1 workflow builder to test changes to individual steps required saving then deploying each change to run a test.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:21]: > In the v1 workflow builder, you had to deploy the whole workflow to test changes to individual steps.

File: docs/docs/migrate-from-v1/README.md [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+We've improved this flow. You can test your changes with a new **Test** button.
+
+In addition to testing single steps, you can now selectively test portions of your workflow in isolation.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:25]: I would provide examples at the end of this sentence:

> you can now selectively test portions of your workflow (e.g. all steps above or below the selected step):

File: docs/docs/migrate-from-v1/README.md [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+
+<img src="./images/test-workflow-portions.png" alt="Selectively pick testing your workflow above or below the current step is now available.">
+
+Now you can also choose to only test steps up to a certain step, or after a current step.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:29]: Then you can remove this sentence after adding the above parenthetical

File: docs/docs/migrate-from-v1/README.md [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+
+### Deploying Changes
+
+After you're happy with your changes, you can **deploy** them to your production workflow. Just click the **Deploy** button in the top right hand corner of the screen.
+
+After deploying your changes, your workflow is now live.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:35]: Maybe add a bit to the end of this last sentence to talk about what this means:

> your workflow is now live, and any changes you made will run against incoming events.

File: docs/docs/migrate-from-v1/README.md [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+
+```javascript
+defineComponent({
+  async run({ steps, $ }) {
+    // your code can be entered here
+  }
+});
+```
+
+1. The `event` from the trigger step is still available, but within `steps.trigger.event` instead.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:61]: "but within `steps.trigger.event` instead" -> "but exposed in `steps.trigger.event` instead.

File: docs/docs/migrate-from-v1/README.md [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+2. The `$` variable has been passed into the `run` function where your code is executed.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:62]: Since this is the first time most people will be seeing the `$`, can you link to the component docs on `$` and mention what this means for people? e.g. that you use `$` method to issue HTTP responses, etc.

File: docs/docs/migrate-from-v1/README.md [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+
+### using 3rd party packages
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:64]: Capitalize "using"

File: docs/docs/migrate-from-v1/README.md [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+import axios from 'axios';
+
+defineComponent({
+  async run({ steps, $ }) {
+    // your code can be entered here
+  }
+});
+```
+
+Allowing all of the scaffolding to be edited opens up the ability to [pass props](code/nodejs/#passing-props-to-code-steps) into your Node.js code steps, which we'll cover later.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:87]: I would recommend moving this to the section where you first reference `defineComponent`. That way you can talk about why we provide the component API and what the benefits are right when the user first learns what it is.

File: docs/docs/migrate-from-v1/README.md [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+
+```javascript
+// this step's name is get_customer_data
+async (event, steps) {
+  this.name = 'Dylan';
+  // downstream steps could use steps.get_customer_data.name to retrieve 'Dylan'
+}
+```
+
+In v2, this specific way of exporting data is now available with `$.export`:
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:101]: "In v2, you use `$.export` to export data, instead:"

File: docs/docs/migrate-from-v1/README.md [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+// this step's name is get_customer_data
+defineComponent({
+  async run({ steps, $ }) {
+    $.export('name', 'Dylan');
+    // downstream steps can use steps.get_customer_data.name to retrieve 'Dylan'
+  }
+});
+```
+::: tip
+Using `return` to export data is the same from v1 to v2. You can still `return` data, and it will be available to other steps with `steps.[stepName].$return_value.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:113]: "is the same in both v1 to v2"

File: docs/docs/migrate-from-v1/README.md [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+:::
+
+### Exiting a workflow early 
+
+In v1, the `$exit` function can be called to exit a flow early:
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:118]: `$exit` -> `$end`

File: docs/docs/migrate-from-v1/README.md [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+
+```javascript
+async (event, steps) {
+  exit('Exiting the whole workflow early');
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:122]: `$end`

File: docs/docs/migrate-from-v1/README.md [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+  console.log('I will never run');
+}
+```
+
+In v2, this same function is available, but under `$.exit`:
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:127]: `$.flow.exit`

File: docs/docs/migrate-from-v1/README.md [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+
+```javascript
+defineComponent({
+  async run({ steps, $ }) {
+    $.flow.exit('Exiting the workflow early');
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:132]: You need to actually `return $.flow.exit` for this to work

File: docs/docs/migrate-from-v1/README.md [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+    console.log('I will never run');
+  }
+});
+```
+
+::: warning
+In v2, `$end` does not display the reason in the workflow results logs.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:139]: There are a few outstanding things like this, but since this should be fixed soon, I would remove this.

File: docs/docs/migrate-from-v1/README.md [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+::: 
+
+### Params vs Props
+
+In workflow builder v2, both Component **props** and Node.js code steps **props** are one in the same.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:144]: I would reframe slightly:

> In the v1 builder, you could pass input to steps using [params](https://pipedream.com/docs/v1/workflows/steps/params/). In the v2 builder, you pass input using [props](/components/props).

File: docs/docs/migrate-from-v1/README.md [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+
+You can still enter free text and select data from other steps in pre-built actions. Also can add your own custom props that accept input like strings, numbers and more just like in v1.
+
+#### Defining params
+
+In the v1 workflow builder, params could be structured or unstructured. The [params schema builder](https://pipedream.com/docs/v1/workflows/steps/params/#configuring-custom-params) allowed you to add your own custom params to steps.
+
+In v2, you can add your own custom props without leaving the code editor with `props`.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:152]: "with `props`" is redundant

File: docs/docs/migrate-from-v1/README.md [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+    }
+  },
+  async run({ steps, $ }) {
+    console.log(this.firstName);
+  }
+});
+
+```
+
+In the example above a `firstName` string prop is created. The value assigned to this prop in the workflow builder by either a static string or dynamic data will be available in the step as `this.firstName`.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:169]: "In the example, you added a `firstName` string prop. "

I would also shorten this sentence, it was hard for me to follow: "The value assigned to this prop in the workflow builder by either a static string or dynamic data will be available in the step as `this.firstName`."

File: docs/docs/migrate-from-v1/README.md [hunk 25]
```diff
@@ -0,0 +1,259 @@
...
+
+Additionally, a visual component is rendered in the step **Configuration** to accept this input:
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:171]: 'Additionally, Pipedream renders a visual component..."

File: docs/docs/migrate-from-v1/README.md [hunk 26]
```diff
@@ -0,0 +1,259 @@
...
+    }
+  },
+  async run({ steps, $ }) {
+    // now your Slack token for authenticating API calls will be available here:
+    this.slack.$auth.oauth_access_token;
+  },
+})
+```
+
+After testing the step the Slack app will appear in the **Configuration** section on the left hand side. In this section you can choose which Slack account you'd like to use in the step.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:196]: "After testing the step, you'll see the Slack app appear..."

Comment (dylburger) [reply docs/docs/migrate-from-v1/README.md:196]: comma after "In this section" in the second sentence

File: docs/docs/migrate-from-v1/README.md [hunk 27]
```diff
@@ -0,0 +1,259 @@
...
+export default defineComponent({
+  props: {
+    twitter: {
+      type: "app",
+      app: "slack",
+    }
+  },
+  async run({ steps, $ }) {
+    // now your Slack token for authenticating API calls will be available here:
+    this.slack.$auth.oauth_access_token;
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:191]: I would recommend modifying this code since I don't think it will work as written. Generally I wouldn't `console.log` it since it's technically credentials, but maybe import the Slack SDK and use the access token to instantiate a Slack client.

File: docs/docs/migrate-from-v1/README.md [hunk 28]
```diff
@@ -0,0 +1,259 @@
...
+```
+
+After testing the step the Slack app will appear in the **Configuration** section on the left hand side. In this section you can choose which Slack account you'd like to use in the step.
+
+<img src="./images/app-props-example.png" alt="Example of adding an app connection to a v2 Node.js step">
+
+
+### HTTPs Response
+
+You can still return an HTTP response from an HTTP triggered workflow.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:203]: "HTTP-triggered"

File: docs/docs/migrate-from-v1/README.md [hunk 29]
```diff
@@ -0,0 +1,259 @@
...
+
+Use `$.respond` to send a JSON or string response from the HTTP call that triggered the workflow.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:205]: Can you link to the `$.respond` docs here? Technically you can return other types of responses (e.g. a Buffer), which we include in those docs.

File: docs/docs/migrate-from-v1/README.md [hunk 30]
```diff
@@ -0,0 +1,259 @@
...
+
+At this time, sharing is not yet implemented in v2 of the workflow builder.
+
+We're working on bringing this same feature to the new version, if you need assistance transferring workflows across accounts, [please contact us](/docs/support).
+
+### `$checkpoint`
+
+The `$checkpoint` functionality to save data between workflow runs has been removed.
+
+But you can leverage the `$.database` service to store arbitrary data across your workflow runs like unique IDs. [Read more about using `$.database` here.](https://pipedream.com/docs/code/nodejs/#managing-state)
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:239]: `$.database` -> `$.service.db`

File: docs/docs/migrate-from-v1/README.md [hunk 31]
```diff
@@ -0,0 +1,259 @@
...
+})
+```
+
+Please note, you'll also need to configure the HTTP trigger step to also allow custom responses. Use the dropdown in the **HTTP Response** section of the HTTP trigger to select the **Return a custom response from your workflow** option.
+
+<img src="./images/custom-http-response-option.png" alt="Select the option to allow your workflow to send it's own HTTP responses">
+
+## Known Gaps & Limitations
+
+The new workflow builder introduces new features that weren't available in v1. However there are some features from the original builder that are not available in v2 currently.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:227]: "However, some features from the original builder are not currently available in v2. The Pipedream team is working to quickly address these items, but if you have feedback that isn't listed here, please [reach out](https://pipedream.com/support)".

File: docs/docs/migrate-from-v1/README.md [hunk 32]
```diff
@@ -0,0 +1,259 @@
...
+
+### Sharing workflows
+
+At this time, sharing is not yet implemented in v2 of the workflow builder.
+
+We're working on bringing this same feature to the new version, if you need assistance transferring workflows across accounts, [please contact us](/docs/support).
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:233]: "We're working on bringing this same feature to the new version. If ..."

File: docs/docs/migrate-from-v1/README.md [hunk 33]
```diff
@@ -0,0 +1,259 @@
...
+
+::: warning
+Please note that any values stored in `$.database` are only accessible in subsequent workflow runs _in the same step_.
+:::
+
+### Public workflows
+
+At this time, all v2 workflows are private.
+
+Unlike v1, it's not possible to make a workflow public. We can still help support your workflow, you just need to share it with us under the workflow settings.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:249]: "We can still help support your workflow, you just need to share it with us under the workflow settings." -> "If you're working with Pipedream support to troubleshoot your workflow, you can share it with the support team under your workflow's Settings"

File: docs/docs/migrate-from-v1/README.md (hunk 12 above)
Comment (ctrlaltdylan) [reply docs/docs/migrate-from-v1/README.md:62]: Of course I go over this in my video recording but not in the text

**Review by dylburger — CHANGES_REQUESTED** [review]
Overall structure / content looks good. I added some specific suggestions inline

File: docs/docs/migrate-from-v1/README.md (hunk 14 above)
Comment (ctrlaltdylan) [reply docs/docs/migrate-from-v1/README.md:87]: I really think passing dynamic props is a bit more advanced than sending HTTP requests.

I updated the sentence to hit that we'll talk about passing props with a link, but I don't want to dive into the more advanced topics right off the bat.

> These changes unify workflow development to the [Component API](/docs/components/) used by pre-built actions and also allows the [defining of props](#defining-props) from within your code steps.

File: docs/docs/migrate-from-v1/README.md (hunk 6 above)
Comment (ctrlaltdylan) [reply docs/docs/migrate-from-v1/README.md:30]: I included a short screencast to reference the hamburger menu to switch contexts. Talking about entering, leaving and re-entering the Inspector seems a little redundant. Read through the newest version and see if that feels right.

File: docs/docs/migrate-from-v1/README.md (hunk 27 above)
Comment (ctrlaltdylan) [reply docs/docs/migrate-from-v1/README.md]: Yea agreed, even though it's something I myself do, we can't say to do it publicly.

The path looks off but it's correct, I just added the exact example from our default action.

**Comment by ctrlaltdylan** [comment]
@dylburger updated with latest edits, thanks for the look over.

I added the FAQ section as well with what I assumed our policies are regarding migrations, end of v1 support, etc.

File: docs/docs/migrate-from-v1/README.md (hunk 14 above)
Comment (dylburger) [reply docs/docs/migrate-from-v1/README.md:102]: nice that works for me

File: docs/docs/migrate-from-v1/README.md [hunk 34]
```diff
@@ -0,0 +1,319 @@
+# Migrate from Pipedream v1
+
+The newest version of the Pipedream workflow builder introduces features like creating code steps in [Python](/code/python), [Bash](/code/bash) and [Go](/code/go).
+
+It also brings the [Node.js](/code/nodes) steps closer to our [Component API](/components/) which powers sources & actions in the platform.
+We have re-imagined the UX from the ground up, made the product much easier to use and have improved performance. In addition, we are introducing powerful new features including:
+
+* **Edit & test** your workflows in separate editing mode without impacting live workflows
+* **Support for multiple languages** including [Node.js](/code/nodejs), [Python](/code/python), [Bash](/code/bash) and [Go](/code/go)
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:9]: You can remove this if you keep the first sentence of this section

File: docs/docs/migrate-from-v1/README.md [hunk 35]
```diff
@@ -0,0 +1,319 @@
+# Migrate from Pipedream v1
+
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:2]: I would start this doc with the first sentence of Tod's note, as well:

> We are excited to announce that we have launched a new version (v2) of Pipedream to all new and existing users!

File: docs/docs/migrate-from-v1/README.md [hunk 36]
```diff
@@ -0,0 +1,319 @@
...
+The newest version of the Pipedream workflow builder introduces features like creating code steps in [Python](/code/python), [Bash](/code/bash) and [Go](/code/go).
+
+It also brings the [Node.js](/code/nodes) steps closer to our [Component API](/components/) which powers sources & actions in the platform.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:5]: Move this first sentence into the paragraph above with multi-lang

File: docs/docs/migrate-from-v1/README.md [hunk 37]
```diff
@@ -0,0 +1,319 @@
+# Migrate from Pipedream v1
+
+The newest version of the Pipedream workflow builder introduces features like creating code steps in [Python](/code/python), [Bash](/code/bash) and [Go](/code/go).
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:3]: Consider shortening to:

> The new builder introduces support for three new languages: ...

File: docs/docs/migrate-from-v1/README.md [hunk 38]
```diff
@@ -0,0 +1,319 @@
...
+
+It also brings the [Node.js](/code/nodes) steps closer to our [Component API](/components/) which powers sources & actions in the platform.
+We have re-imagined the UX from the ground up, made the product much easier to use and have improved performance. In addition, we are introducing powerful new features including:
+
+* **Edit & test** your workflows in separate editing mode without impacting live workflows
+* **Support for multiple languages** including [Node.js](/code/nodejs), [Python](/code/python), [Bash](/code/bash) and [Go](/code/go)
+* **Granular testing** including the ability to test individual steps and more
+* **Multiple triggers** are now supported per workflow
+* **Improved** forms for easier configuration and streamlined building
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:12]: Keep **forms** in bold, as well

File: docs/docs/migrate-from-v1/README.md [hunk 39]
```diff
@@ -0,0 +1,319 @@
+# Migrate from Pipedream v1
+
+The newest version of the Pipedream workflow builder introduces features like creating code steps in [Python](/code/python), [Bash](/code/bash) and [Go](/code/go).
+
+It also brings the [Node.js](/code/nodes) steps closer to our [Component API](/components/) which powers sources & actions in the platform.
+We have re-imagined the UX from the ground up, made the product much easier to use and have improved performance. In addition, we are introducing powerful new features including:
+
+* **Edit & test** your workflows in separate editing mode without impacting live workflows
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:8]: "Edit and test" -> ampersands should only be used in formal titles, etc. (Dreyer's style guide)

File: docs/docs/migrate-from-v1/README.md [hunk 40]
```diff
@@ -0,0 +1,319 @@
...
+
+### Editing vs Inspecting
+
+In v1, editing your workflow and inspecting past events were visible in the same view. The new v2 builder has improved this by separating the workflow **Builder** from the workflow events **Inspector**.
+
+Switch between these contexts using the menu in the top right of the workflow builder.
+
+<img src="./images/new-builder-context-switcher.gif" alt="Switching between the Builder & Inspector contexts in the v2 workflow builder">
+
+When you first open a deployed workflow, you're presented with the **Inspector** version of the workflow. In this view you can see logs of past events, and select them to see the results of each step in the workflow.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:28]: **Inspector** version -> **Inspector** _version_

File: docs/docs/migrate-from-v1/README.md (hunk 6 above)
Comment (dylburger) [reply docs/docs/migrate-from-v1/README.md:30]: Makes sense and the GIF looks good

File: docs/docs/migrate-from-v1/README.md [hunk 41]
```diff
@@ -0,0 +1,319 @@
...
+
+To edit the workflow, click the **Edit** button in the top right hand corner. This will close the inspector and allow you to edit your workflow without the distraction of logs from the production flow.
+
+### Testing Changes
+
+In the v1 workflow builder, you had to deploy the whole workflow to test changes to individual steps. This meant you were
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:34]: "This meant you were" -> incomplete sentence

File: docs/docs/migrate-from-v1/README.md [hunk 42]
```diff
@@ -0,0 +1,319 @@
...
+
+There are a few changes to the Node.js code steps that you should know about. Some functions have been renamed for more clarity, and we've aligned the Node.js code steps closer to the [Component API](/docs/components/).
+
+### Code Scaffolding Format
+
+In v1, the Node.js steps would automatically scaffold new Node.js steps in this format:
+
+```javascript
+async (event, steps) {
+  // your code could be entered in here
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:58]: super small nit but you can shorten this to "your code here"

File: docs/docs/migrate-from-v1/README.md [hunk 43]
```diff
@@ -0,0 +1,319 @@
...
+// this step's name is get_customer_data
+defineComponent({
+  async run({ steps, $ }) {
+    $.export('name', 'Dylan');
+    // downstream steps can use steps.get_customer_data.name to retrieve 'Dylan'
+  }
+});
+```
+::: tip
+Using `return` to export data is the same from v1 to v2. You can still `return` data, and it will be available to other steps with `steps.[stepName].$return_value.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:128]: Need a trailing backtick at the end of this variable

File: docs/docs/migrate-from-v1/README.md [hunk 44]
```diff
@@ -0,0 +1,319 @@
...
+      app: "slack",
+    }
+  },
+  async run({ steps, $ }) {
+    return await axios($, {
+      url: `https://slack.com/api/users.profile.get`,
+      headers: {
+        Authorization: `Bearer ${this.slack.$auth.oauth_access_token}`,
+      },
+    }
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:208]: Needs a closing `)`

File: docs/docs/migrate-from-v1/README.md [hunk 45]
```diff
@@ -0,0 +1,319 @@
...
+After testing the step, you'll see the Slack app will appear in the **Configuration** section on the left hand side. In this section you can choose which Slack account you'd like to use in the step.
+
+<img src="./images/app-props-example.png" alt="Example of adding an app connection to a v2 Node.js step">
+
+
+### HTTP Response
+
+You can still return an HTTP response from an HTTP-triggered workflow.
+
+Use [`$.respond`](/docs/workflows/steps/triggers/#http) to send a JSON or string response from the HTTP call that triggered the workflow.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:222]: Link goes to HTTP trigger section instead of `$.respond`

File: docs/docs/migrate-from-v1/README.md [hunk 46]
```diff
@@ -0,0 +1,319 @@
...
+
+In v2, you can test and save your progress on a workflow _without_ deploying it.
+
+However, after deploying it's not possible to rollback to a prior version.
+
+## FAQs
+
+### What are the benefits of the new (v2) workflow builder?
+
+* **Edit & test** your workflows in separate editing mode without impacting live workflows
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:278]: Same comments from the top section here

File: docs/docs/migrate-from-v1/README.md [hunk 47]
```diff
@@ -0,0 +1,319 @@
...
+* **Support for multiple languages** including Node, Python, Golang & bash
+* **Granular testing** including the ability to test individual steps and more
+* **Multiple triggers** are now supported per workflow
+* **Improved** forms for easier configuration and streamlined building
+
+### What are the limitations of the new (v2) workflow builder?
+
+* `$checkpoint` has been removed from v2 workflows, but `$.database` is a near equivalent.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:286]: `$.service.db`

Comment (dylburger) [reply docs/docs/migrate-from-v1/README.md:286]: "`$.database` is a near equivalent" -> "`$.service.db` provides a similar API"

File: docs/docs/migrate-from-v1/README.md [hunk 48]
```diff
@@ -0,0 +1,319 @@
...
+* Sharing workflows is not supported
+* Making workflows public is not supported
+* Workflows are no longer versioned
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:289]: I would say "You cannot roll a workflow back to a previous version" since workflows are technically still versioned, and we should be able to support rollback once the UI supports.

File: docs/docs/migrate-from-v1/README.md [hunk 49]
```diff
@@ -0,0 +1,319 @@
...
+
+[The limitations of the new workflow builder are covered in detail here](#limitations).
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:291]: To change to active voice, you can say something like "You can read more about these limitations here"

File: docs/docs/migrate-from-v1/README.md [hunk 50]
```diff
@@ -0,0 +1,319 @@
...
+
+### Are v2 workflows backwards compatible?
+
+If your workflow is not using custom Node.js code and is compromised of only prebuilt actions, then there's a good chance it's supported.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:295]: Since people don't really know the distinction between v1 and v2 actions, I don't think we should say this. We're still a bit of a ways out to get consistent v2 action coverage.

Maybe you can state the facts of the situation plainly for users and encourage them to reach out if they observe specific gaps, e.g. you can say that we only support component actions in the new builder and that we're working quickly to migrate all actions over. If you notice a missing action, please reach out.

File: docs/docs/migrate-from-v1/README.md [hunk 51]
```diff
@@ -0,0 +1,319 @@
...
+### What are the limitations of the new (v2) workflow builder?
+
+* `$checkpoint` has been removed from v2 workflows, but `$.database` is a near equivalent.
+* Sharing workflows is not supported
+* Making workflows public is not supported
+* Workflows are no longer versioned
+
+[The limitations of the new workflow builder are covered in detail here](#limitations).
+
+### Are v2 workflows backwards compatible?
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:293]: After writing the below comment, I think we should just remove this section completely. You cover all of the important information in "How do I migrate v1 workflows to v2 workflows?".

File: docs/docs/migrate-from-v1/README.md [hunk 52]
```diff
@@ -0,0 +1,358 @@
...
+
+### Rolling back a specific version
+
+In v2, you can test and save your progress on a workflow _without_ deploying it.
+
+However, after deploying it's not possible to rollback to a prior version of a deployed workflow.
+
+You can still edit a deployed workflow, just like in v1 but automatic version rollbacks are not currently possible.
+
+## FAQs
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:305]: Re: the conversation standup, we should have an FAQ for something like "Where did my list of events on the left go?"

File: docs/docs/migrate-from-v1/README.md [hunk 53]
```diff
@@ -0,0 +1,358 @@
...
+
+### How do I migrate v1 workflows to v2 workflows?
+
+At this time we do not have an automated process to change v1 to v2. To create a v2 equivalent workflow, you can recompose your v1 workflow in the v2 builder.
+
+However, if it uses custom Node.js code steps, be sure to [follow the changes we describe in the guide above](/migrate-from-v1/#node-js-code-step-changes).
+
+### When will the new (v2) workflow builder be the default builder for all customers?
+
+v2 is currently an opt in feature, you can still create and edit v1 workflows if you have an grandfathered account.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:348]: I would change this to section to read:

By default, existing users will still default to the v1 builder. You can create new v2 workflows from the dropdown menu next to the **New** workflow button.

if you'd like to default to the v2 builder when creating new workflows, you can change the **Builder Version** in [your account settings](https://pipedream.com/settings/account).

File: docs/docs/migrate-from-v1/README.md [hunk 54]
```diff
@@ -0,0 +1,358 @@
...
+
+We will not make the v2 builder the default workflow experience until we have reached feature parity with v1 by providing paths to overcome the [limitations of v2](#known-gaps-limitations).
+
+### When will I no longer be able to create v1 workflows?
+
+There is currently no deprecation date for v1 workflows, we will continue to support of v1 workflows until we have feature parity with v2.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:371]: Change comma to period here:

> There is currently no deprecation date for v1 workflows. We ...

File: docs/docs/migrate-from-v1/README.md [hunk 55]
```diff
@@ -0,0 +1,358 @@
...
+
+When this date becomes clear we will provide assistance to automatically and assist migrate v1 to v2 workflows for you.
```
Comment (dylburger) [inline docs/docs/migrate-from-v1/README.md:373]: Since we have no specific plan for automated migration yet, I would remove this section, and we can address that later.

**Review by dylburger — CHANGES_REQUESTED** [review]
See suggestions inline

File: docs/docs/migrate-from-v1/README.md (hunk 52 above)
Comment (ctrlaltdylan) [reply docs/docs/migrate-from-v1/README.md:328]: Added this 👍

File: docs/docs/migrate-from-v1/README.md (hunk 53 above)
Comment (ctrlaltdylan) [reply docs/docs/migrate-from-v1/README.md:371]: Right, forgot we added that setting good call out

File: docs/docs/migrate-from-v1/README.md (hunk 48 above)
Comment (ctrlaltdylan) [reply docs/docs/migrate-from-v1/README.md:343]: Got it. You should make a comment on the email / Slack annoucement too about this, I copy & pasted from there

Comment (dylburger) [reply docs/docs/migrate-from-v1/README.md:343]: ack I'll let Tod know
